File: Chap1_Arrays&Strings/1.3_URLify/URLify.js
/**
 * @param {string} str
 * @param {number} n
 * @returns {string}
 */
function URLify(str, n) {
    let slow = n - 1;
    let fast = n - 1;
    let tail = true;
    let arr = str.split("");
    while (slow > 0 && fast > 0) {
        while (tail && arr[fast] == " ") {
            fast--;
        }
        tail = false;
        if (arr[fast] !== " ") {
            arr[slow--] = arr[fast--];
        } else {
            arr[slow--] = "0";
            arr[slow--] = "2";
            arr[slow--] = "%";
            fast--;
        }
    }
    return arr.join("");
}

function test() {
    let str = "my name is tony      ";
    console.log(URLify(str, str.length));
}

File: Chap1_Arrays&Strings/1.3_URLify/URLify.java
class Main {
    public static void main(String[] args) {
        char[] arr = "Hi Tony Stark    ".toCharArray();
        System.out.println(URLify(arr, arr.length));
    }

    static String URLify(char[] arr, int n) {
        // use two pointers
        // the fast one is for finding chars
        // the slow one is for filling in chars and %20
        int slow = n - 1;
        int fast = n - 1;
        boolean tail = true;
        while (slow > 0 && fast > 0) {
            // try to find the last char of the original string
            while (tail && arr[fast] == ' ') {
                fast--;
            }
            tail = false;
            if (arr[fast] != ' ') {
                arr[slow--] = arr[fast--];
            } else {
                arr[slow--] = '0';
                arr[slow--] = '2';
                arr[slow--] = '%';
                fast--;
            }
        }
        return String.valueOf(arr);
    }
}